### PR TITLE
Refactor: rewrite `Merge` method to address readability and efficiency

### PR DESCRIPTION
**Pull request by chendave**
Signed-off-by: Dave Chen <dave.chen@arm.com>

@ahg-g @lixiang233 I came across this line of code today, and I do think this line of code is a legacy nit, the status code has been set in the below snippet of code, but in the loop it was set for **each plugin**, consider the case that status map hold the status for each of plugin, this kind of instruction is not effective and unnecessary.

https://github.com/kubernetes/kubernetes/blob/c5cc25d1cb075f84756db9bffecd3fafe12c5fab/pkg/scheduler/framework/interface.go#L202-L208

For the sake of other status code might be added in the future, we can tweak the code a  little bit.

@ahg-g  what do you think?

**What type of PR is this?**

Add one of the following kinds:
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by alculquicondor** [comment]
I don't think this PR is increasing readability.

However, I do think this function is hard to read. I would reconsider if you can come up with an alternate rewrite of the function that is more readable.

**Comment by ahg-g** [comment]
right, this doesn't move the readability needle.

**Comment by chendave** [comment]
@alculquicondor  How about this version?

File: pkg/scheduler/framework/interface.go [hunk 1]
```diff
@@ -184,28 +184,35 @@ func (p PluginToStatus) Merge() *Status {
...
 	finalStatus := NewStatus(Success)
-	var hasUnschedulableAndUnresolvable, hasUnschedulable bool
 	for _, s := range p {
-		if s.Code() == Error {
-			finalStatus.err = s.AsError()
-		} else if s.Code() == UnschedulableAndUnresolvable {
-			hasUnschedulableAndUnresolvable = true
-		} else if s.Code() == Unschedulable {
-			hasUnschedulable = true
+		switch finalStatus.code {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:188]: I think the most readable approach is to map statuses to precedence numbers. And then replace finalStatus.code if the precedence of s.Status is higher.

Comment (chendave) [reply pkg/scheduler/framework/interface.go]: yes, that's better.

File: pkg/scheduler/framework/interface.go [hunk 2]
```diff
@@ -183,29 +183,28 @@ func (p PluginToStatus) Merge() *Status {
 		return nil
 	}
 
+	statusPrecedenceMap := map[Code]int{
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:186]: move this outside the function. And nit: just `statusPrecedence` is fine

**Review by alculquicondor — COMMENTED** [review]
Don't forget to update title and description

File: pkg/scheduler/framework/interface.go (hunk 2 above)
Comment (chendave) [reply pkg/scheduler/framework/interface.go]: Done

**Comment by chendave** [comment]
/retitle "Refactor: rewrite `Merge` method to address readability and efficiency"

**Comment by chendave** [comment]
/retitle Refactor: rewrite Merge method to address readability and efficiency

File: pkg/scheduler/framework/interface.go [hunk 3]
```diff
@@ -74,6 +74,15 @@ const (
 	Skip
 )
 
+// StatusPrecedence defines a map from status to its precedence, larger value means higher precedent.
+var StatusPrecedence = map[Code]int{
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:78]: Do not export this

Comment (chendave) [reply pkg/scheduler/framework/interface.go:78]: Done

File: pkg/scheduler/framework/interface.go [hunk 4]
```diff
@@ -74,6 +74,15 @@ const (
 	Skip
 )
 
+// statusPrecedence defines a map from status to its precedence, larger value means higher precedent.
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:77]: sorry, last nit: have this after `var codes` (because it references the enum).

**Review by alculquicondor — COMMENTED** [review]
/approve

File: pkg/scheduler/framework/interface.go (hunk 4 above)
Comment (chendave) [reply pkg/scheduler/framework/interface.go:80]: Good catch, done

**Pull request merged**